# Strings, List Dictionary

In this module you'll dive into more advanced ways to manipulate strings using indexing, slicing, and advanced formatting. You'll also explore the more advanced data types: lists, tuples, and dictionaries. You'll learn to store, reference, and manipulate data in these structures, as well as combine them to store complex data structures.




**Key Concepts**
- Manipulate strings using indexing, slicing, and formatting
- Use lists and tuples to store, reference, and manipulate data
- Leverage dictionaries to store more complex data, reference data by keys, and manipulate data stored
- Combine these data types to construct complex data structures



# Strings

Basic Structures Introduction 


We have covered the basic elements of Python syntax. We talked about how to define functions, how to make your computer act differently based on conditionals, and how to make it perform operations repeatedly using while, and for loops, and recursion. Now that we have the basics of syntax out of the way, we can start growing our Python knowledge which will let us do more and more interesting operations. Remember, one of our main goals in this course is to help you learn to write short Python scripts that automate actions, you've made big steps towards getting there.

**Data Type/Structure:**
- Strings
- Lists
- Dictionaries


Modify the double_word function so that it returns the same word repeated twice, followed by the length of the new doubled word. For example, double_word("hello") should return hellohello10.

In [158]:
def double_word(word):
    result = word * 2
    length = len(result)
    result += str(length)
    return result 

print(double_word("hello")) # Should return hellohello10
print(double_word("abc"))   # Should return abcabc6
print(double_word(""))      # Should return 0

hellohello10
abcabc6
0


Next we'll cover some of the operations we can perform over strings, including how to access parts of them and modify them.

In [25]:
name = "Jaylen"
print(name[1])
print(name[0])
print(name[5])
print(name[6])

a
J
n


IndexError: string index out of range

What if you want to print the last character of a string but you don't know how long it is? 

In this example, we don't know the length of the string, but it doesn't matter. Using negative indexes lets us access the positions in the string starting from the last

In [30]:
text = "Random string  with a lot of charachtors"
print(text[-1])
print(text[-5])



s
h


False

Want to give it a go yourself? Be my guest! Modify the first_and_last function so that it returns True if the first letter of the string is the same as the last letter of the string, False if they’re different. Remember that you can access characters using message[0] or message[-1]. Be careful how you handle the empty string, which should return True since nothing is equal to nothing.

In [55]:
def first_and_last(message):
    if message!="":
        if (message[0] == message[-1]):
            return True
        else:
            return False
    else:
        return True
print(first_and_last("else"))
print(first_and_last("tree"))
print(first_and_last(""))

True
False
True


In [160]:
def first_and_last(message):
    if message == "" or message[0] == message[-1]:
        return True
    return False

print(first_and_last("else"))
print(first_and_last("tree"))
print(first_and_last(""))

True
False
True


On top of accessing individual characters, we can also access a slice of a string. A slice is the portion of a string that can contain more than one character, also sometimes called a substring. We do that by creating a range using a colon as a separator. Let's see an example of this.


In [65]:
color = "orange"
color[0:4]

'oran'

In [61]:
fruit = "pineapple"
print(fruit[:4])

'pine'

In [64]:
print(fruit[4:])

apple


Modify and Creating New Strings


In [72]:
message = "A kong string with a silly typo" 
message[2]= 'l'

TypeError: 'str' object does not support item assignment

These pesky type errors, right? In this case, we're told that strings don't support item assignment. This means that we can't change individual characters because strings in Python are immutable, which is just a fancy word, meaning they can't be modified. What we can do is create a new string based on the old one, like this.

In [73]:
new_message = message[0:2] + 'l' + message[3:]
print(new_message)

A long string with a silly typo


 But does this mean the message variable can never change? Not really. We can assign a new value to the same variable. Let's do that a couple of times to see how it works.

In [75]:
message = "this is the message"
print(message)

this is the message


In [78]:
message = "And another message"
print(message)

And another message


So, we figured out how to create a new message from the old one. But how are we supposed to know which character to change? Let's try something different.


In [83]:
pets= "Cats & Dogs"
pets.index("&")

5

In this case, we're using a method to get the index of a certain character. A method is a function associated with a specific class. We'll talk a lot more about classes and methods later. 

 For now, what you need to know is that this is a function that applies to a variable. And we can call it by following the variable with a dot. Let's try this a few more times.

In [85]:
pets.index("C")

0

In [86]:
pets.index("Dog")

7

In [88]:
pets.index("s")

3

Here, we know there are two s characters, but we only get one value. That's because the index method returns just the first position that matches. And what happens if the string doesn't have the substring we're looking for? The index method can't return a number because the substring isn't there, so we get a value error instead.

In [89]:
pets.index("S")

ValueError: substring not found

Try using the index method yourself now!

Using the index method, find out the position of "x" in "supercalifragilisticexpialidocious".

In [92]:
word = "supercalifragilisticexpialidocious"
print(word.index("x"))

21


We said, that if the substring isn't there, we would get an error. So how can we know if a substring is contained in a string to avoid the error? Let's check this out

- by using `in` keyword

In [94]:
"Drogon" in pets

False

In [96]:
"Cat" in pets

True

Let's put all the stuff together to solve a real-world problem. Imagine that your company has recently moved to using a new domain, but a lot of the company email addresses are still using the old one. You want to write a program that replaces this old domain with the new one in any outdated email addresses.

The function to replace the domain would look like this.

In [164]:
def replace_domain(email, old_domain, new_domain):
    if "@" + old_domain in email:
        index = email.index("@" + old_domain)
        new_email = email[:index] + "@" + new_domain
        return new_email
    return email
    
# replace_domain(morteza@me.com, me.com, me2.com)

# More string methods
 let's get back to the good stuff. So far, we've seen ways you can access portions of strings using the indexing technique, create new strings by slicing and concatenating, find characters and strings using the index method, and even test if one string contains another. On top of all this string processing power, the string class provides a bunch of other methods for working with text. Now, we'll show you how to use some of these methods. Remember, the goal is not for you to memorize all of this. Instead, we want to give you an idea of what you can do with strings in Python.

- Transformation or farmating on string 
    - Handeling user input using 
        - .upper()
        - .lower()
        

In [104]:
"montain".upper()

'MONTAIN'

In [105]:
"MOUTAIN".lower()

'moutain'

Let's say you wanted to check if the user answered yes to a question. How would you know if the user typed it using upper or lower case? You don't need to, you just transform the answer to the case you want. Like this example.

In [108]:
answer = "YES"
if answer.lower()=="yes":
    print("User said yes")

User said yes


# strip method 
Another useful method when dealing with user input is the strip method. This method will get rid of surrounding spaces in the string. If we ask the user for an answer, we usually don't care about any surrounding spaces. So it's a good idea to use the strip method to get rid of any white space.

This means that strip doesn't just remove spaces, it also removes tabs and new line characters, which are all characters we don't usually want in user-provided strings. 

- lstrip
- rstrip





In [112]:
" yes ".strip()

'yes'

In [113]:
" yes ".lstrip()

'yes '

In [115]:
" yes ".rstrip()

' yes'

In [118]:
"the number of times e occurs in this string is 4".count("e")

4

The method endswith returns whether the string ends with a certain substring.

In [121]:
"Forest".endswith("rest")

True

In [128]:
"123".isnumeric()

True

In [127]:
int("123") + int("321")

444

In earlier videos, we showed that we can concatenate strings using the plus sign. The join method can also be used for concatenating.

In [131]:
" ".join(["This", "is", "a" , "phrase", "joined", "by", "space"])

'This is a phrase joined by space'

In [132]:
"...".join(["This", "is", "a" , "phrase", "joined", "by", "space"])

'This...is...a...phrase...joined...by...space'

Finally, we can also split a string into a list of strings.


In [135]:
"This is an example".split()

['This', 'is', 'an', 'example']

Want to try some string methods yourself? Give it a go!

Fill in the gaps in the initials function so that it returns the initials of the words contained in the phrase received, in upper case. For example: "Universal Serial Bus" should return "USB"; "local area network" should return "LAN”.

In [169]:
def initials(phrase):
    words = phrase.split()
    result = ""
    for word in words:
        result += word[0].upper()
    return result
print(initials("Universal Serial Bus")) # Should be: USB
print(initials("local area network")) # Should be: LAN
print(initials("Operating system")) # Should be: OS

USB
LAN
OS


In [166]:
"Universal Serial Bus".split()

['Universal', 'Serial', 'Bus']

We've covered a bunch of String class methods already, so let's keep building on those and run down some more advanced methods.

The string method lower will return the string with all characters changed to lowercase. The inverse of this is the upper method, which will return the string all in uppercase. Just like with previous methods, we call these on a string using dot notation, like "this is a string".upper(). This would return the string "THIS IS A STRING". This can be super handy when checking user input, since someone might type in all lowercase, all uppercase, or even a mixture of cases.

You can use the strip method to remove surrounding whitespace from a string. Whitespace includes spaces, tabs, and newline characters. You can also use the methods  lstrip and rstrip to remove whitespace only from the left or the right side of the string, respectively.

The method count can be used to return the number of times a substring appears in a string. This can be handy for finding out how many characters appear in a string, or counting the number of times a certain word appears in a sentence or paragraph.

If you wanted to check if a string ends with a given substring, you can use the method endswith. This will return True if the substring is found at the end of the string, and False if not.

The isnumeric method can check if a string is composed of only numbers. If the string contains only numbers, this method will return True. We can use this to check if a string contains numbers before passing the string to the int() function to convert it to an integer, avoiding an error. Useful!

We took a look at string concatenation using the plus sign, earlier. We can also use the join method to concatenate strings. This method is called on a string that will be used to join a list of strings. The method takes a list of strings to be joined as a parameter, and returns a new string composed of each of the strings from our list joined using the initial string. For example, " ".join(["This","is","a","sentence"]) would return the string "This is a sentence".

The inverse of the join method is the split method. This allows us to split a string into a list of strings. By default, it splits by any whitespace characters. You can also split by any other characters by passing a parameter.

#### Formatting Strings

 Up to now we've been making strings using the plus sign to just concatenate the parts of the string we wanted to create. And we've used the str function to convert numbers into strings so that we can concatenate them, too. This works, but it's not ideal, especially when the operations you want to do with the string or on the tricky side. There's a better way to do this using the format method. Let's see a couple of examples.
 
 - Format Method
 
#### Format options enables you to output humane readable information that you ask for. It is very reach.
 

In [181]:
name = "Manny"
number = len(name) * 3
print("Hello {},  your lucky number is {} ".format(name, number))

Hello Manny,  your lucky number is 15 


In [196]:
print("your lucky number is {number}, {name}.".format(name=name, number=len(name)*3))

your lucky number is 15, Manny.


Modify the student_grade function using the format method, so that it returns the phrase "X
received Y% on the exam". For example, student_grade("Reed", 80) should return "Reed received 80% on the exam".

In [187]:
def student_grade(name, grade):
    return "{name} received {grade}% on the exam".format(grade=grade, name=name)
print(student_grade("Reed", 80))
print(student_grade("Paige", 92))
print(student_grade("Jesse", 85))

Reed received 80% on the exam
Paige received 92% on the exam
Jesse received 85% on the exam


In [819]:
price = 7.5
with_tax = price * 1.09
print(price, with_tax)

7.5 8.175


et's say you want to output the price of an item with and without tax. Depending on what the tax rate is, the number might be a long number with a bunch of decimals. So if something costs `$7.5` without tax and the tax rate is 9%, the price with tax would be $8.175. First off, ouch, and also, since there's no such thing as half a penny anymore, that number doesn't make sense. So to fix this we can make the format function print only two decimals, like this.

In [822]:
print("Base price: ${:.2f}. with Tax: ${:.2f}".format(price, with_tax))

Base price: $7.50. with Tax: $8.18


In [195]:
def to_celcius(x):
    return (x-32) * 5/9
for x in range(0, 101, 10):
    print ("{:>3} F | {:>6.2f} C".format(x, to_celcius(x)))

  0 F | -17.78 C
 10 F | -12.22 C
 20 F |  -6.67 C
 30 F |  -1.11 C
 40 F |   4.44 C
 50 F |  10.00 C
 60 F |  15.56 C
 70 F |  21.11 C
 80 F |  26.67 C
 90 F |  32.22 C
100 F |  37.78 C


In the first expression we're saying we want the numbers to be aligned to the right for a total of three spaces. In the second expression we're saying we want the number to always have exactly two decimal places and we want to align it to the right at six spaces. We can use string formatting like this to make the output of our program look nice and also to generate useful logging and debugging messages.


# Summary format() Method
You can use the **format** method on strings to concatenate and format strings in all kinds of powerful ways. To do this, create a string containing curly brackets, **{}**, as a placeholder, to be replaced. Then call the format method on the string using .format() and pass variables as parameters. The variables passed to the method will then be used to replace the curly bracket placeholders. This method automatically handles any conversion between data types for us. 

If the curly brackets are empty, they’ll be populated with the variables passed in the order in which they're passed. However, you can put certain expressions inside the curly brackets to do even more powerful string formatting operations. You can put the name of a variable into the curly brackets, then use the names in the parameters. This allows for more easily readable code, and for more flexibility with the order of variables.

You can also put a formatting expression inside the curly brackets, which lets you alter the way the string is formatted. For example, the formatting expression **{:.2f}** means that you’d format this as a float number, with two digits after the decimal dot. The colon acts as a separator from the field name, if you had specified one. You can also specify text alignment using the greater than operator: >. For example, the expression **{:>3.2f}** would align the text three spaces to the right, as well as specify a float number with two decimal places. String formatting can be very handy for outputting easy-to-read textual output.

# String Reference Cheet Sheet 

In Python, there are a lot of things you can do with strings. In this cheat sheet, you’ll find the most common string operations and string methods.

**String operations**
- len(string) returns the length of the string 
- for character in string Iterates over each character in the string
- if substring in string Checks whether the substring is part of the string
- `string[i]` Accesses the character at index i of the string, starting at zero
- `string[i:j]` Accesses the substring starting at index i, ending at index j-1. If i is omitted, it's 0 by default. If j is omitted, it's len(string) by default.

**String methods** 
- `string.lower()` / `string.upper()` Returns a copy of the string with all lower / upper case characters 
- `string.lstrip()` / `string.rstrip()` / `string.strip()` Returns a copy of the string without left / right / left or right whitespace
-string.count(substring) Returns the number of times substring is present in the string
- `string.isnumeric()` Returns True if there are only numeric characters in the string. If not, returns False.
- `string.isalpha()` Returns True if there are only alphabetic characters in the string. If not, returns False.
- `string.split()` / string.split(delimiter) Returns a list of substrings that were separated by whitespace / delimiter
- `string.replace(old, new)` Returns a new string where all occurrences of old have been replaced by new.
- `delimiter.join(list of strings)` Returns a new string with all the strings joined by the delimiter


Check out the official documentation for [all available String methods](https://docs.python.org/3/library/stdtypes.html#string-methods).

# Formatting Strings Cheat Sheet
Python offers different ways to format strings. In the video, we explained the format() method. In this reading, we'll highlight three different ways of formatting strings. For this course you only need to know the format() method. But on the internet, you might find any of the three, so it's a good idea to know that the others exist.



#### Using the format() method
The format method returns a copy of the string where the {} placeholders have been replaced with the values of the variables. These variables are converted to strings if they weren't strings already. Empty placeholders are replaced by the variables passed to format in the same order.

In [202]:
# "base string with {} placeholders".format(variables)

example = "format() method"

formatted_string = "this is an example of using the {} on a string".format(example)

print(formatted_string)



this is an example of using the format() method on a string


If the placeholders indicate a number, they’re replaced by the variable corresponding to that order (starting at zero).



In [206]:
# "{0} {1}".format(first, second)

first = "apple"
second = "banana"
third = "carrot"

formatted_string = "{0} {2} {1}".format(first, second, third)

print(formatted_string)


apple carrot banana


If the placeholders indicate a field name, they’re replaced by the variable corresponding to that field name. This means that parameters to format need to be passed indicating the field name.



In [208]:
# "{var1} {var2}".format(var1=value1, var2=value2)

If the placeholders include a colon, what comes after the colon is a formatting expression. See below for the expression reference.

Official documentation for the format [string syntax](https://docs.python.org/3/library/string.html#formatstrings)

In [ ]:
# {:d} integer value
'{:d}'.format(10.5) → '10'


#### Formatting expressions

![](Img\Formatting_expressions.PNG)



Check out the official documentation for [all available expressions](https://docs.python.org/3/library/string.html#format-specification-mini-language).


#### Old string formatting (Optional)
The format() method was introduced in Python 2.6. Before that, the % (modulo) operator could be used to get a similar result. While this method is no longer recommended for new code, you might come across it in someone else's code. This is what it looks like:

"base string with %s placeholder" % variable

The % (modulo) operator returns a copy of the string where the placeholders indicated by %  followed by a formatting expression are replaced by the variables after the operator.

"base string with %d and %d placeholders" % (value1, value2)

To replace more than one value, the values need to be written between parentheses. The formatting expression needs to match the value type.

-----
"%(var1) %(var2)" % {var1:value1, var2:value2}


Variables can be replaced by name using a dictionary syntax (we’ll learn about dictionaries in an upcoming video).

"Item: %s - Amount: %d - Price: %.2f" % (item, amount, price)

The formatting expressions are mostly the same as those of the format() method. 

Check out the official documentation for [old string formatting](https://docs.python.org/3/library/stdtypes.html#old-string-formatting).


#### Formatted string literals (Optional)

This feature was added in Python 3.6 and isn’t used a lot yet. Again, it's included here in case you run into it in the future, but it's not needed for this or any upcoming courses.

A formatted string literal or f-string is a string that starts with 'f' or 'F' before the quotes. These strings might contain {} placeholders using expressions like the ones used for format method strings.

The important difference with the format method is that it takes the value of the variables from the current context, instead of taking the values from parameters.

Examples:
> name = "Micah"

> print(f'Hello {name}')

Hello Micah

----------

> item = "Purple Cup"

> amount = 5

> price = amount * 3.25

> print(f'Item: {item} - Amount: {amount} - Price: {price:.2f}')

Item: Purple Cup - Amount: 5 - Price: 16.25

Check out the official documentation for [f-strings](https://docs.python.org/3/reference/lexical_analysis.html#f-strings).

# Practice Quiz: Strings

# Q1
The is_palindrome function checks if a string is a palindrome. A palindrome is a string that can be equally read from left to right or right to left, omitting blank spaces, and ignoring capitalization. Examples of palindromes are words like kayak and radar, and phrases like "Never Odd or Even". Fill in the blanks in this function to return True if the passed string is a palindrome, False if not.




In [697]:
#EXAMPLE 6: 互文
def is_palindrome(input_string):
	# We'll create two strings, to compare them
	new_string = ""
	reverse_string = ""
	# Traverse through each letter of the input string
	for x in input_string.upper():
		# Add any non-blank letters to the 
		# end of one string, and to the front
		# of the other string. 
		if x != " ":
			new_string = new_string + x
			reverse_string = x + reverse_string
	# Compare the strings
	if new_string == reverse_string:
		return True
	return False

print(is_palindrome("Never Odd or Even")) # Should be True
print(is_palindrome("abc")) # Should be False
print(is_palindrome("kayak")) # Should be True


True
False
True


# Q2
Using the format method, fill in the gaps in the convert_distance function so that it returns the phrase "X miles equals Y km", with Y having only 1 decimal place. For example, convert_distance(12) should return "12 miles equals 19.2 km".


In [246]:
def convert_distance(miles):
	km = miles * 1.6 
	result = "{} miles equals {:.2f} km".format(miles, km)
	return result

print(convert_distance(12)) # Should be: 12 miles equals 19.2 km
print(convert_distance(5.5)) # Should be: 5.5 miles equals 8.8 km
print(convert_distance(11)) # Should be: 11 miles equals 17.6 km

12 miles equals 19.20 km
5.5 miles equals 8.80 km
11 miles equals 17.60 km


# Q3

If we have a string variable named Weather = "Rainfall", which of the following will print the substring or all characters before the "f"?


`print(Weather[:4])`

`print(Weather[4:])`

`print(Weather[1:4])`

`print(Weather[:"f"])`






In [224]:
weather = 'Rainfall'
print(weather[:4])

Rain


# Q4
Fill in the gaps in the nametag function so that it uses the format method to return first_name and the first initial of last_name followed by a period. For example, nametag("Jane", "Smith") should return "Jane S."



In [227]:
def nametag(first_name, last_name): 
    return("{} {}.".format(first_name, last_name[0]))
print(nametag("Jane", "Smith")) 
# Should display "Jane S." 
print(nametag("Francesco", "Rinaldi")) 
# Should display "Francesco R." 
print(nametag("Jean-Luc", "Grand-Pierre")) 
# Should display "Jean-Luc G." 

Jane S.
Francesco R.
Jean-Luc G.


# Q5
The replace_ending function replaces the old string in a sentence with the new string, but only if the sentence ends with the old string. If there is more than one occurrence of the old string in the sentence, only the one at the end is replaced, not all of them. For example, replace_ending("abcabc", "abc", "xyz") should return abcxyz, not xyzxyz or xyzabc. The string comparison is case-sensitive, so replace_ending("abcabc", "ABC", "xyz") should return abcabc (no changes made).



In [249]:
def replace_ending(sentence, old, new):
	# Check if the old string is at the end of the sentence 
	if sentence.endswith(old):
		# Using i as the slicing index, combine the part
		# of the sentence up to the matched string at the 
		# end with the new string
		i = len(old) # sentence.index(old)
		new_sentence = sentence[:-i] + new 
		return new_sentence

	# Return the original sentence if there is no match 
	return sentence
	
print(replace_ending("It's raining cats and cats", "cats", "dogs")) 
# Should display "It's raining cats and dogs"
print(replace_ending("She sells seashells by the seashore", "seashells", "donuts")) 
# Should display "She sells seashells by the seashore"
print(replace_ending("The weather is nice in May", "may", "april")) 
# Should display "The weather is nice in May"
print(replace_ending("The weather is nice in May", "May", "April")) 
# Should display "The weather is nice in April"

It's raining cats and dogs
She sells seashells by the seashore
The weather is nice in May
The weather is nice in April


# Lists

#####  What is a list?




In [252]:
 x = ["Now", "we", "are", "cooking"]
type(x)

list

In [254]:
print(x)

['Now', 'we', 'are', 'cooking']


In [255]:
len (x)

4

To check if a list contains a certain element, you can use the keyword "in" like in these examples.

In [256]:
"are" in x  

True

Again, like when we use this with strings, the result of this check is a Boolean, which we can use as a condition for branching or looping.

In [257]:
print(x[0])

Now


In [258]:
print(x[3])

cooking


In [260]:
print(x[4])

IndexError: list index out of range

We get an index error. We can't go over the end of the list. Remember that because list indexes start at zero, accessing the item at index four means we're trying to access the fifth element in the list. There are only four elements. So we're out of range if we try to access the index number four.



![](Img\indexing_list.PNG)


 Create slice of a list just like doing it with string

In [264]:
x[1:3]  # 3-1

['we', 'are']

We can also leave out one of the range indexes empty. The first value defaults to zero and the second value to the length of the list

In [283]:
x[2:]

['are', 'cooking']

Using the "split" string method from the preceding lesson, complete the get_word function to return the {n}th word from a passed sentence. For example, get_word("This is a lesson about lists", 4) should return "lesson", which is the 4th word in this sentence. Hint: remember that list indexes start at 0, not 1.

In [291]:
def get_word(sentence, n):
	# Only proceed if n is positive 
	if n > 0:
		words = sentence.split()
		# Only proceed if n is not more than the number of words 
		if n <= len(words):
			return(words[n-1])
	return("")
print(get_word("This is a lesson about lists", 4)) # Should print: lesson
print(get_word("This is a lesson about lists", -4)) # Nothing
print(get_word("Now we are cooking!", 1)) # Should print: Now
print(get_word("Now we are cooking!", 5)) # Nothing

lesson

Now



In Python, strings and lists are both examples of **sequences**. There are other sequences too, and they all share a bunch of operations like

- iterating over them using for-loops
- indexing using the len function to know the length of the sequence,
- using plus to concatenate two sequences
- using in to verify if the sequence contains an element.

#### Modifying the Contents of a List

- One of the ways that lists and strings are different is that lists are mutable. Which is another fancy word to say that they can change. 

- append method
     - Adding an element to a list using the 
     
 The element always gets added to the end. You could start with an empty list and add all of its items using append

In [315]:
fruits = ["Pineapple", "Banna", "Apple", "Melone"]
fruits.append("Kiwi")
print(fruits)

['Pineapple', 'Banna', 'Apple', 'Melone', 'Kiwi']


- insert method

Adds the element to the give index

In [316]:
fruits.insert(0, "Orangr ")
print(fruits)

['Orangr ', 'Pineapple', 'Banna', 'Apple', 'Melone', 'Kiwi']


What happens if we use a number larger than the length of the list?

In [317]:
fruits.insert(25, 'peach')
print(fruits)

['Orangr ', 'Pineapple', 'Banna', 'Apple', 'Melone', 'Kiwi', 'peach']


- Remove Method

In [318]:
fruits.remove("Melone")
print(fruits)

['Orangr ', 'Pineapple', 'Banna', 'Apple', 'Kiwi', 'peach']


- pop method
   - Another way we can remove elements is by using the pop method, which receives an index
   - he pop method returns the element that was removed at the index that was passed

In [319]:
fruits.pop(3)

'Apple'

In [320]:
print(fruits)

['Orangr ', 'Pineapple', 'Banna', 'Kiwi', 'peach']


In the last way to modify the contents of a list is to change an item by assigning something else to that position, like this.

In [322]:
fruits[2]="Strawberry"
print(fruits)

['Orangr ', 'Pineapple', 'Strawberry', 'Kiwi', 'peach']


Modifying the contents of lists will come up in tons of scripts as we operate with them. If the list contains hosts on a network, you could add or remove hosts as they come online or offline. If the list contains users authorized to run a certain process, you could add or remove users when permissions are granted or removed and so on.

The `skip_elements` function returns a list containing every other element from an input list, starting with the first element. Complete this function to do that, using the for loop to iterate through the input list.

In [60]:
def skip_elements(elements):
    new_list = []
    for i in range(0, len(elements), 2):
        new_list.append(elements[i])
    return (new_list)
print(skip_elements(["a", "b", "c", "d", "e", "f", "g"])) # Should be ['a', 'c', 'e', 'g']
print(skip_elements(['Orange', 'Pineapple', 'Strawberry', 'Kiwi', 'Peach'])) # Should be ['Orange', 'Strawberry', 'Peach']
print(skip_elements([])) # Should be []

['a', 'c', 'e', 'g']
['Orange', 'Strawberry', 'Peach']
[]


In [65]:
def skip_elements(elements):
    new_list = []
    i = 0
    for element in elements:
        if element == elements[i]:
            new_list.append(elements[i])
            i +=2
    return new_list
print(skip_elements(["a", "b", "c", "d", "e", "f", "g"])) # Should be ['a', 'c', 'e', 'g']
print(skip_elements(['Orange', 'Pineapple', 'Strawberry', 'Kiwi', 'Peach'])) # Should be ['Orange', 'Strawberry', 'Peach']
print(skip_elements([])) # Should be []

['a', 'c', 'e', 'g']
['Orange', 'Strawberry', 'Peach']
[]


Whenever you need to write a program that'll handle a variable amount of elements, you'll use a list. What if you need a sequence of a fixed amount of elements? That comes in the following content.

## Summary: Modify the content of a list
While lists and strings are both sequences, a big difference between them is that lists are mutable. This means that the contents of the list can be changed, unlike strings, which are immutable. You can add, remove, or modify elements in a list.

You can add elements to the end of a list using the append method. You call this method on a list using dot notation, and pass in the element to be added as a parameter. For example, list.append("New data") would add the string "New data" to the end of the list called list.

If you want to add an element to a list in a specific position, you can use the method insert. The method takes two parameters: the first specifies the index in the list, and the second is the element to be added to the list. So list.insert(0, "New data") would add the string "New data" to the front of the list. This wouldn't overwrite the existing element at the start of the list. It would just shift all the other elements by one. If you specify an index that’s larger than the length of the list, the element will simply be added to the end of the list.

You can remove elements from the list using the remove method. This method takes an element as a parameter, and removes the first occurrence of the element. If the element isn’t found in the list, you’ll get a ValueError error explaining that the element was not found in the list.

You can also remove elements from a list using the pop method. This method differs from the remove method in that it takes an index as a parameter, and returns the element that was removed. This can be useful if you don't know what the value is, but you know where it’s located. This can also be useful when you need to access the data and also want to remove it from the list.

Finally, you can change an element in a list by using indexing to overwrite the value stored at the specified index. For example, you can enter list[0] = "Old data" to overwrite the first element in a list with the new string "Old data".

### Lists and Tuples

**there are a number of data types in Python that are all sequences.**
- Strings are sequences of characters and are **immutable**.
- Lists are sequences of elements of any type and are **mutable**.
- A third data type that's a sequence and also closely related to lists is the tuple.


You might be wondering, why do we even need another sequence type? Weren't lists great? Yes, lists are great. They can hold any number of elements and we can add, remove and modify their contents as much as we want, but **there are cases when we want to make sure an element in a certain position or index refers to one specific thing and won't change**. In these situations, lists won't help us. Thanks for nothing lists. 

In the following example, we have a tuple that represents someone's full name. The **first element of the tuple is the first-name**. **The second element is the middle initial**, and **the third element is the last-name.**

In [66]:
fullame = ('Grace' , 'M', 'Hopper')

**Tuples usage**

 Tuples are used for lots of different things in Python. One common example is the return value of functions. When a function returns more than one value, it's actually returning a tuple. Remember the function to convert seconds to hours, minutes, and seconds that we saw a while back? Here just to remind you. This function returns three values. In other words, it returns a tuple of three elements. Let's give it a try.
 
 The out put has three elements as we expected. A tuple with three elements
1. hours
2. muints
3. seconds

In [73]:
def convert_seconds(seconds):
    hours = seconds // 3600
    minuts = (seconds - hours * 3600)// 60
    remainig_seconds= seconds - hours * 3600 - minuts * 60 
    return hours, minuts, remainig_seconds
result = convert_seconds(5000)
print(result)

(1, 23, 20)


**unpack tuples**


This means that we can turn a tuple of three elements into three separate variables. Because the order won't change, we know what those variables are present, like this.

In [74]:
hours, minuts, seconds = result 
print(hours, minuts, seconds)

1 23 20


So now we've split the tuple into three separate values. We've seen before that we can also do this directly when calling the function without the intermediate result variable.

In [75]:
hours, minuts, seconds = convert_seconds(5000) 
print(hours, minuts, seconds)

1 23 20


In Python, it's really common to use tuples to represent data that has more than one value and that needs to be kept together. For example, you could use a tuple to have a filename and it's size, or you could store the name and email address of a person, or a date and time and the general health of the system at any point in time. Can you see how these different data types could help you automate some of your IT work?

Let's use tuples to store information about a file: its name, its type and its size in bytes. Fill in the gaps in this code to return the size in kilobytes (a kilobyte is 1024 bytes) up to 2 decimal places.

In [86]:
def file_size(file_info):
    name, type, size = file_info
    return("{:.2f}".format(size/1024))

print(file_size(('Class Assignment', 'docx', 17875))) # Should print 17.46
print(file_size(('Notes', 'txt', 496))) # Should print 0.48
print(file_size(('Program', 'py', 1239))) # Should print 1.21

17.46
0.48
1.21


Knowing when to use tuples and when to use lists can seem a little fuzzy at first, but don't worry, it'll get clearer as we tackle more examples.

#### Iterating over Lists and Tuples
for calculating


In [102]:
animals = ["Lion", "Zebra", "Dolphin", "Monkey"]
chars = 0
for animal in animals:
    chars+=len(animal)
print("Total Characters: {}, Avrange length: {:.1f}".format(chars, chars/len(animals)))


Total Characters: 22, Avrange length: 5.5


What if you wanted to know the index of an element while going through the list? You could use the range function and then use indexing to access the elements at the index that range returned. You could use a range function and then use indexing to access the elements at the index that range just returned or you could just use the **enumerate function**. 

In [103]:
winners = ['Ashley', 'Dylan', 'Reese']
for index, persone in enumerate(winners):
    print('{}-{}'.format(index + 1, persone))

1-Ashley
2-Dylan
3-Reese


Try out the enumerate function for yourself in this quick exercise. Complete the skip_elements function to return every other element from the list, this time using the enumerate function to check if an element is on an even position or an odd position.

In [135]:
new_list = []
for index, element in enumerate(["a", "b", "c", "d", "e", "f", "g"]):
    new_list = index, element
    print(new_list)

(0, 'a')
(1, 'b')
(2, 'c')
(3, 'd')
(4, 'e')
(5, 'f')
(6, 'g')


In [119]:
def skip_elements(elements):
    new_elements = []
    for index, element in enumerate(elements):
        if index % 2 == 0:
            new_elements.append(element)
    return new_elements
print(skip_elements(["a", "b", "c", "d", "e", "f", "g"])) # Should be ['a', 'c', 'e', 'g']
print(skip_elements(['Orange', 'Pineapple', 'Strawberry', 'Kiwi', 'Peach'])) # Should be ['Orange', 'Strawberry', 'Peach']
print(skip_elements([])) # Should be []

['a', 'c', 'e', 'g']
['Orange', 'Strawberry', 'Peach']
[]


Write a function that creates a new list containing one string per person including their name and the email address between angled brackets. the format usually used in emails like this. So what do we need to do?

In [128]:
def full_emails(people):
    result = []
    for email, name in people:
        result.append("{} <{}>".format(name, email))
    return result
people = [("morteza@example.com", "Morteza Gomes"),("shay@example.com", "shay Brandt")]
print(full_emails(people))

['Morteza Gomes <morteza@example.com>', 'shay Brandt <shay@example.com>']


ommon errors when dealing with lists in Python. Because we use the range function so much with for loops, you might be tempted to use it for iterating over indexes of a list and then to access the elements through indexing. You could be particularly inclined to do this if you're used to other programming languages before. Because in some languages, the only way to access an element of a list is by using indexes. 


Real talk, this works but looks ugly. 

It's more idiomatic in Python to iterate through the elements of the list directly or using enumerate when you need the indexes like we've done so far. 

There are some specific cases that do require us to iterate over the indexes, for example, when we're trying to modify the elements of the list we're iterating. 


By the way, if you're iterating through a list and you want to modify it at the same time, you need to be very careful. If you remove elements from the list while iterating, you're likely to end up with an unexpected result. In this case, it might be better to use a copy of the list instead. We've now seen a bunch of different things we can do with lists, and hopefully you're starting to see how they can be a very powerful tool in your IT specialist toolkit.

**Next up, we're going to learn a powerful technique for creating lists.**


### Summary

When we covered for loops, we showed the example of iterating over a list. This lets you iterate over each element in the list, exposing the element to the for loop as a variable. But what if you want to access the elements in a list, along with the index of the element in question? You can do this using the `enumerate()` function. The enumerate() function takes a list as a parameter and returns a tuple for each element in the list. The first value of the tuple is the index and the second value is the element itself.

### List Comprehension (creating list in a short way)

Say we wanted to create a list with multiples of 7 from 7 to 70, we could do it like this


In [142]:
multiple = []
for x in range(1, 11):
    multiple.append(x*7)
print(multiple)

[7, 14, 21, 28, 35, 42, 49, 56, 63, 70]


This works fine and is a good way of solving it. But because creating lists based on sequences is such a common task Python provides a technique called list comprehension, that lets us do it in just one line. This is how we would do the same with list comprehension. 

In [143]:
multiple = [x*7 for x in range(1, 11)]
print(multiple)

[7, 14, 21, 28, 35, 42, 49, 56, 63, 70]


**List comprehensions**

let us create new lists based on **sequences or ranges**. So we can use this technique whenever we want to create a list based on a range like in this example. Or based on the contents of a list a tuple a string or any other Python sequence.

Generate a list of the length of the strings inside the list called language.

In [147]:
languages = ["Python", "Perl", "Ruby", "Go", "Java", "C"]
length = [len(language) for language in languages]
print(length)

[6, 4, 4, 2, 4, 1]


List comprehensions also let us use a conditional clause. Say we wanted all the numbers that are divisible by 3 between 0 and a 100, we could create a list like this

In [152]:
z = [x for x in range(0, 101) if x % 3 == 0]
print(z)

[0, 3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48, 51, 54, 57, 60, 63, 66, 69, 72, 75, 78, 81, 84, 87, 90, 93, 96, 99]


The odd_numbers function returns a list of odd numbers between 1 and n, inclusively. Fill in the blanks in the function, using list comprehension. Hint: remember that list and range counters start at 0 and end at the limit minus 1.



In [158]:
def odd_numbers(n):
    odd_list = [x for x in range(1, n+1) if x % 2 != 0]
    return odd_list
print(odd_numbers(5))  # Should print [1, 3, 5]
print(odd_numbers(10)) # Should print [1, 3, 5, 7, 9]
print(odd_numbers(11)) # Should print [1, 3, 5, 7, 9, 11]
print(odd_numbers(1))  # Should print [1]
print(odd_numbers(-1)) # Should print []

[1, 3, 5]
[1, 3, 5, 7, 9]
[1, 3, 5, 7, 9, 11]
[1]
[]


 Using list comprehensions when programming in Python is totally optional. Sometimes it can make the code look nicer and more readable, at other times it can have the opposite effect, especially if we try to pack too much information together. In general, it's a good idea to know that list comprehensions exist, especially when you're trying to understand someone else's code. 
 
 
 ### Summaty
 You can create lists from sequences using a for loop, but there’s a more streamlined way to do this: list comprehension. List comprehensions allow you to create a new list from a sequence or a range in a single line.

For example, `[ x*2 for x in range(1,11) ]` is a simple list comprehension. This would iterate over the range 1 to 10, and multiply each element in the range by 2. This would result in a list of the multiples of 2, from 2 to 20.

You can also use conditionals with list comprehensions to build even more complex and powerful statements. You can do this by appending an if statement to the end of the comprehension. For example, `[ x for x in range(1,101) if x % 10 == 0 ]` would generate a list containing all the integers divisible by 10 from 1 to 100. The if statement we added here evaluates each value in the range from 1 to 100 to check if it’s evenly divisible by 10. If it is, it gets added to the list.

List comprehensions can be really powerful, but they can also be super complex, resulting in code that’s hard to read. Be careful when using them, since it might make it more difficult for someone else looking at your code to easily understand what the code is doing.

## Lists and Tuples Operations Cheat Sheet
Lists and tuples are both sequences, so they share a number of sequence operations. But, because lists are mutable, there are also a number of methods specific just to lists. This cheat sheet gives you a run down of the common operations first, and the list-specific operations second.

###### Common sequence operations
- len(sequence) Returns the length of the sequence
- for element in sequence Iterates over each element in the sequence
- if element in sequence Checks whether the element is part of the sequence
- sequence[i] Accesses the element at index i of the sequence, starting at zero
- sequence[i:j] Accesses a slice starting at index i, ending at index j-1. If i is omitted, it's 0 by default. If j is omitted, it's len(sequence) by default.
- for index, element in enumerate(sequence) Iterates over both the indexes and the elements in the sequence at the same time

Check out the official documentation for [sequence operations](https://docs.python.org/3/library/stdtypes.html#sequence-types-list-tuple-range).


###### List-specific operations and methods
- `list[i] = x` Replaces the element at index i with x
- `list.append(x)` Inserts x at the end of the list
- `list.insert(i, x)` Inserts x at index i
- `list.pop(i)` Returns the element a index i, also removing it from the list. If i is omitted, the last element is returned and removed.
- `list.remove(x)` Removes the first occurrence of x in the list
- `ist.sort()` Sorts the items in the list
- `list.reverse()` Reverses the order of items of the list
- `list.clear()` Removes all the items of the list
- `list.copy()` Creates a copy of the list
- `list.extend(other_list)` Appends all the elements of other_list at the end of list

Most of these methods come from the fact that lists are mutable sequences. For more info, see the [official documentation](https://docs.python.org/3/library/stdtypes.html#mutable-sequence-types) for mutable sequences and the [list specific documentation](https://docs.python.org/3/library/stdtypes.html#lists).

###### List comprehension
1. [expression for variable in sequence] Creates a new list based on the given sequence. Each element is the result of the given expression.

2. [expression for variable in sequence if condition] Creates a new list based on the given sequence. Each element is the result of the given expression; elements only get added if the condition is true.


# Practice Quiz: Lists

# Q1 

Given a list of filenames, we want to rename all the files with extension hpp to the extension h. To do this, we would like to generate a new list called newfilenames, consisting of the new filenames. Fill in the blanks in the code using any of the methods you’ve learned thus far, like a for loop or a list comprehension.



In [222]:
filenames = ["program.c", "stdio.hpp", "sample.hpp", "a.out", "math.hpp", "hpp.out"]
# Generate newfilenames as a list containing the new filenames
newfilenames = []
# using as many lines of code as your chosen method requires.
for x in range(len(filenames)):
        old_name = filenames[x]
        if old_name.endswith(".hpp"):
            new_name = old_name.replace('.hpp', '.h')
            newfilenames.append(new_name)
        else:
            newfilenames.append(old_name)
print(newfilenames)
            
        
# Should be ["program.c", "stdio.h", "sample.h", "a.out", "math.h", "hpp.out"]

['program.c', 'stdio.h', 'sample.h', 'a.out', 'math.h', 'hpp.out']


In [227]:
filenames = ["program.c", "stdio.hpp", "sample.hpp", "a.out", "math.hpp", "hpp.out"]
newfilenames = []
for filename in filenames:
    old_name = filename
    if old_name.endswith(".hpp"):
        new_name = old_name.replace('.hpp', '.h')
        newfilenames.append(new_name)
    else:
        newfilenames.append(old_name)
print(newfilenames)

['program.c', 'stdio.h', 'sample.h', 'a.out', 'math.h', 'hpp.out']


In [223]:
filenames = ["program.c", "stdio.hpp", "sample.hpp", "a.out", "math.hpp", "hpp.out"]
test = []
for filename in filenames:
    if "hpp" in filename:
        print(filename)  
    

stdio.hpp
sample.hpp
math.hpp
hpp.out


# Q2
Let's create a function that turns text into pig latin: a simple text transformation that modifies each word moving the first character to the end and appending "ay" to the end. For example, python ends up as ythonpay.


In [230]:
def pig_latin(text):
  say = []
  # Separate the text into words
  words = text.split()
  for word in words:
    # Create the pig latin word and add it to the list
    word  = word[1:] + word[0] + 'ay'
    say.append(word)
    # Turn the list back into a phrase
  return ' '.join(say)
		
print(pig_latin("hello how are you")) # Should be "ellohay owhay reaay ouyay"
print(pig_latin("programming in python is fun")) # Should be "rogrammingpay niay ythonpay siay unfay"

ellohay owhay reaay ouyay
rogrammingpay niay ythonpay siay unfay


In [233]:
def pig_latin(text):
    pigged_text = []
    for word in text.split(" "):
        word = word[1:] + word[0] + "ay"
        pigged_text.append(word)
    return " ".join(pigged_text)
		
print(pig_latin("hello how are you")) # Should be "ellohay owhay reaay ouyay"
print(pig_latin("programming in python is fun")) # Should be "rogrammingpay niay ythonpay siay unfay"

ellohay owhay reaay ouyay
rogrammingpay niay ythonpay siay unfay


# Q3
The permissions of a file in a Linux system are split into three sets of three permissions: read, write, and execute for the owner, group, and others. Each of the three values can be expressed as an octal number summing each permission, with `4 corresponding to read`, `2 to write`, and `1 to execute`. Or it can be written with a string using the letters `r`, `w`, and `x` or `-` when the permission is not granted. For example: `640` is `read/write` for the `owner`, `read` for the `group`, and `no permissions` for the `others`; converted to a string, it would be: `"rw-r-----" `755 is read/write/execute for the owner, and read/execute for group and others; converted to a string, it would be: `"rwxr-xr-x"` Fill in the blanks to make the code convert a permission in octal format into a string format.



In [295]:
def octal_to_string(octal):
    result = ""
    value_letters = [(4,"r"),(2,"w"),(1,"x")]
    # Iterate over each of the digits in octal
    for digit in [int(n) for n in str(octal)]:
        # Check for each of the permissions values
        for value, letter in value_letters:
            if digit >= value:
                result += letter
                digit -= value
            else:
                result+='-'
    return result
    
print(octal_to_string(755)) # Should be rwxr-xr-x
print(octal_to_string(644)) # Should be rw-r--r--
print(octal_to_string(750)) # Should be rwxr-x---
print(octal_to_string(600)) # Should be rw-------

rwxr-xr-x
rw-r--r--
rwxr-x---
rw-------


In [274]:
octal = 755
for digit in [int(n) for n in str(octal)]:
    print(digit)

7
5
5


In [273]:
value_latters = [(4, 'r'), (2, 'w'), (1, 'x')]
for value, latter in value_latters:
    print(value, latter)

4 r
2 w
1 x


# Q4
Tuples and lists are very similar types of sequences. What is the main thing that makes a tuple different from a list?


![](Img\tuplesVSlist.PNG)

# Q5
The group_list function accepts a group name and a list of members, and returns a string with the format: group_name: member1, member2, … For example, `group_list("g", ["a","b","c"])` returns `"g: a, b, c"`. Fill in the gaps in this function to do that.



In [292]:
def group_list(group, users):
    members = ", ".join(users)
    return group + ': ' + members
    
print(group_list("Marketing", ["Mike", "Karen", "Jake", "Tasha"])) # Should be "Marketing: Mike, Karen, Jake, Tasha"
print(group_list("Engineering", ["Kim", "Jay", "Tom"])) # Should be "Engineering: Kim, Jay, Tom"
print(group_list("Users", "")) # Should be "Users:"

Marketing: Mike, Karen, Jake, Tasha
Engineering: Kim, Jay, Tom
Users: 


 # Q5
The guest_list function reads in a list of tuples with the name, age, and profession of each party guest, and prints the sentence `"Guest is X years old and works as __."` for each one. For example, `guest_list(('Ken', 30, "Chef"), ("Pat", 35, 'Lawyer'), ('Amanda', 25, "Engineer"))` should print out: `Ken is 30 years old and works as Chef.` `Pat is 35 years old and works as Lawyer.` `Amanda is 25 years old and works as Engineer.` Fill in the gaps in this function to do that.



In [399]:
def guest_list(guests):
	for guest in guests:
		name, age, role  = guest
		print("{name} is {age} years old and works as {role}".format(name=name, age=age, role=role))

guest_list([('Ken', 30, "Chef"), ("Pat", 35, 'Lawyer'), ('Amanda', 25, "Engineer")])


Ken is 30 years old and works as Chef
Pat is 35 years old and works as Lawyer
Amanda is 25 years old and works as Engineer


# Dictionary

- Like lists, dictionaries are used to organize elements into collections. 
- Unlike lists, you don't access elements inside dictionaries using their position.
- **Instead, the data inside dictionaries take the form of pairs of keys and values**.
- To get a dictionary value we use its corresponding key.
- Another way these two vary is while in a list the index must be a number, in a dictionary you can use a bunch of different data types as keys, like strings, integers, floats, tuples, and more.

- The name dictionaries comes from how they work in a similar way to **human language dictionaries**. In an English language dictionary the **word comes with a definition**.

- **In the language of a Python dictionary, the word would be the key and the definition would be the value.**

In [403]:
x = {}
type(x)

dict

- Creating initialized dictionaries isn't too different from the syntax we used in earlier videos to create initialized lists or tuples.

- But instead of a series of slots with values in them, we have a series of keys that point at values. Okay, let's check out an example dictionary.




In a dictionary, it's perfectly fine to mix and match the data types of keys and values like this and can be very useful. In this example, we're using a dictionary to store the number of files corresponding to each extension. It makes sense to encode the file extension formatting in a string, while it's natural to represent a count as an integer number. Let's say you want to find out how many text files there are in the dictionary. To do this, you would use the key txt to access its associated value. The syntax to do this may look familiar, since we used something similar in our examples of indexing strings, lists, and tuples.

In [406]:
file_counts ={"jpeg":10, "txt":14, "csv":2, "py":23}
print(file_counts)

{'jpeg': 10, 'txt': 14, 'csv': 2, 'py': 23}



In this file_counts dictionary, we've stored keys that are strings, like jpg, that point at integer values, like 10. When creating the dictionary we use colons and between the key and the value and separate each pair by commas. In a dictionary, it's perfectly fine to mix and match the data types of keys and values like this and can be very useful.

In [407]:
file_counts['txt']

14

In [410]:
"jpeg" in file_counts

True

In [411]:
'html' in file_counts

False

Dictinaries are Mutable. 

You can create add, remove and replace entries

In [412]:
file_counts['cfg'] = 8
print(file_counts)

{'jpeg': 10, 'txt': 14, 'csv': 2, 'py': 23, 'cfg': 8}


What if we have the same key name avilable in the dictianry while adding a new entry?
Here, check it out!
This will replace the value for the same key in dictinary

In [413]:
file_counts['csv'] = 10
print(file_counts)

{'jpeg': 10, 'txt': 14, 'csv': 10, 'py': 23, 'cfg': 8}


Last off, we can delete elements from a dictionary with the del keyword by passing the dictionary and the key to the element as if we were trying to access it

In [426]:
del file_counts['cfg']
print(file_counts)

{'jpeg': 10, 'txt': 14, 'csv': 10, 'py': 23}


The "toc" dictionary represents the table of contents for a book. Fill in the blanks to do the following: 

- 1) Add an entry for Epilogue on page 39. 
- 2) Change the page number for Chapter 3 to 24. 
- 3) Display the new dictionary contents. 
- 4) Display True if there is Chapter 5, False if there
isn't.

In [427]:
toc = {"Introduction":1, "Chapter 1":4, "Chapter 2":11, "Chapter 3":25, "Chapter 4":30}
toc['Epilogue']=39 # Epilogue starts on page 39
print(toc)

{'Introduction': 1, 'Chapter 1': 4, 'Chapter 2': 11, 'Chapter 3': 25, 'Chapter 4': 30, 'Epilogue': 39}


In [428]:
toc['Chapter 3']=24 # Chapter 3 now starts on page 24
print(toc)

{'Introduction': 1, 'Chapter 1': 4, 'Chapter 2': 11, 'Chapter 3': 24, 'Chapter 4': 30, 'Epilogue': 39}


In [429]:
print(toc) # What are the current contents of the dictionary?

{'Introduction': 1, 'Chapter 1': 4, 'Chapter 2': 11, 'Chapter 3': 24, 'Chapter 4': 30, 'Epilogue': 39}


In [431]:
if 'chapter 5' in toc:
    print(True)
else:
    print(False)

False


# Sumaries Dictionaries
Dictionaries are another data structure in Python. They’re similar to a list in that they can be used to organize data into collections. However, data in a dictionary isn't accessed based on its position. Data in a dictionary is organized into pairs of keys and values. You use the key to access the corresponding value. Where a list index is always a number, a dictionary key can be a different data type, like a string, integer, float, or even tuples.

When creating a dictionary, you use curly brackets: `{}`. When storing values in a dictionary, the key is specified first, followed by the corresponding value, separated by a colon. For example, `animals = { "bears":10, "lions":1, "tigers":2 }` creates a dictionary with three key value pairs, stored in the variable animals. The key "bears" points to the integer value 10, while the key "lions" points to the integer value 1, and "tigers" points to the integer 2. You can access the values by referencing the key, like this: `animals["bears"]`. This would return the integer 10, since that’s the corresponding value for this key.

You can also check if a key is contained in a dictionary using the in keyword. Just like other uses of this keyword, it will return True if the key is found in the dictionary; otherwise it will return False.

Dictionaries are mutable, meaning they can be modified by adding, removing, and replacing elements in a dictionary, similar to lists. You can add a new key value pair to a dictionary by assigning a value to the key, like this: `animals["zebras"] = 2`. This creates the new key in the animal dictionary called zebras, and stores the value 2. You can modify the value of an existing key by doing the same thing. So `animals["bears"] = 11` would change the value stored in the bears key from 10 to 11. Lastly, you can remove elements from a dictionary by using the del keyword. By doing `del animals["lions"]` you would remove the key value pair from the animals dictionary.

#### Iterating over the Contents of a Dictionary


In [433]:
file_counts ={"jpeg":10, "txt":14, "csv":2, "py":23}
for extention in file_counts:
    print(extention)

jpeg
txt
csv
py


- Use `items` to get key value pairs
- `keys` to get the keys
- `values` to get just the values.

In [439]:
for ext, amount in file_counts.items():
    print("There are {} files with the .{} extention".format(amount, ext))

There are 10 files with the .jpeg extention
There are 14 files with the .txt extention
There are 2 files with the .csv extention
There are 23 files with the .py extention


- Access to value or keys

In [440]:
file_counts.values()

dict_values([10, 14, 2, 23])

In [443]:
file_counts.keys()

dict_keys(['jpeg', 'txt', 'csv', 'py'])

These methods return special data types related to the dictionary, but you don't need to worry about what they are exactly. You just need to iterate them as you would with any sequence

In [447]:
for value in file_counts.values():
    print(value)

10
14
2
23


Now, it's your turn! Have a go at iterating over a dictionary!

Complete the code to iterate through the keys and values of the cool_beasts dictionary. Remember that the items method returns a tuple of key, value for each element in the dictionary.



In [449]:
cool_beasts = {"octopuses":"tentacles", "dolphins":"fins", "rhinos":"horns"}
for key, value in cool_beasts.items():
    print("{} have {}".format(key, value))

octopuses have tentacles
dolphins have fins
rhinos have horns


Because we know that each key can be present only once, dictionaries are a great tool for counting elements and analyzing frequency. Let's check out a simple example of counting how many times each letter appears in a piece of text.

In [829]:
def count_letters(text):
    result = {}
    for letter in text:
        if letter not in result:
            result[letter] = 0 
        result[letter] += 1
    return result
count_letters("aaaaa")

{'a': 5}

In [452]:
count_letters("tenant")

{'t': 2, 'e': 1, 'n': 2, 'a': 1}

In [455]:
print(count_letters("a long string with a long letters"), end= ' ')

{'a': 2, ' ': 6, 'l': 3, 'o': 2, 'n': 3, 'g': 3, 's': 2, 't': 4, 'r': 2, 'i': 2, 'w': 1, 'h': 1, 'e': 2} 

To only count the letters, we'd need to specify which characters we're taking into account. This technique might seem simple at first, but it can be really useful in a lot of cases. Let's say for example that you're analyzing logs in your server and you want to count how many times each type of error appears in the log file. You could easily do this with a dictionary by using the type of error as the key and then incrementing the associated value each time you come across that error type. Are you starting to see how dictionaries can be a really useful tool when writing scripts? 

 #### Coming up, we're going to learn how to tell when to use dictionaries and when to use lists.

#### Dictionaries vs. Lists


In [459]:
ip_address = ["192.168.1.1", "127.0.0.1", "8.8.8.8"]
host_address = {"router": "192.168.1.1", "localhost":"127.0.0.1", "google":"127.0.0.1"}
for i in host_address.items():
    print(i)

('router', '192.168.1.1')
('localhost', '127.0.0.1')
('google', '127.0.0.1')


In [468]:
wardrobe = {"shirt":["red","blue","white"], "jeans":["blue","black"]}
for item in wardrobe.keys():
    print(item)

shirt
jeans


In Python, a dictionary can only hold a single value for a given key. To workaround this, our single value can be a list containing multiple values. Here we have a dictionary called "wardrobe" with items of clothing and their colors. Fill in the blanks to print a line for each item of clothing with each color, for example: "red shirt", "blue shirt", and so on.

In [471]:
wardrobe = {"shirt":["red","blue","white"], "jeans":["blue","black"]}
for item, colors in wardrobe.items():
    for color in colors:
        print("{} {}".format(color, item))

red shirt
blue shirt
white shirt
blue jeans
black jeans


There are even more data types available that we haven't checked out yet. 

- One of these data types is a set which is a bit like a cross between a list and a dictionary.

- A set is used when you want to store a bunch of elements and be certain that there are only present once. 

- Elements of a set must also be immutable.

- You can think of this as the keys of a dictionary with no associated values or you could see it as a list where what matters isn't the order of the elements but whether an element is in the list or not.

### Dictionary Methods Cheat Sheet

- Definition
`x = {key1:value1, key2:value2}`

- Operations
     - `len(dictionary)` Returns the number of items in the dictionary
     - `for key in dictionary` Iterates over each key in the dictionary
     - for key, value in `dictionary.items()` - Iterates over each key,value pair in the dictionary
     - if key in dictionary - Checks whether the key is in the dictionary
     - `dictionary[key]` - Accesses the item with key key of the dictionary
     - `dictionary[key] = value` - Sets the value associated with key
     - `del dictionary[key]` - Removes the item with key key from the dictionary

     
     
- Methods
     - `dict.get(key, default)` - Returns the element corresponding to key, or default if it's not present
     - `dict.keys()` - Returns a sequence containing the keys in the dictionary
     - `dict.values()` - Returns a sequence containing the values in the dictionary
     - `dict.update(other_dictionary)` - Updates the dictionary with the items coming from the other dictionary. Existing entries will be replaced; new entries will be added.
     - `dict.clear()` - Removes all the items of the dictionary
     
Check out the [official documentation for dictionary operations and methods](https://docs.python.org/3/library/stdtypes.html#mapping-types-dict).




# Practice Quiz: Dictionaries
# Q1
The email_list function receives a dictionary, which contains domain names as keys, and a list of users as values. Fill in the blanks to generate a list that contains complete email addresses (e.g. diana.prince@gmail.com).



In [511]:
def email_list(dictionary):
    emails = []
    for domain,users in dictionary.items():
        for user in users:
            emails.append(user + "@" + domain)
    return emails        
print(email_list({"gmail.com": ["clark.kent", "diana.prince", "peter.parker"], 
                  "yahoo.com": ["barbara.gordon", "jean.grey"], "hotmail.com": ["bruce.wayne"]}))

['clark.kent@gmail.com', 'diana.prince@gmail.com', 'peter.parker@gmail.com', 'barbara.gordon@yahoo.com', 'jean.grey@yahoo.com', 'bruce.wayne@hotmail.com']


In [510]:
dic = {"gmail.com": ["clark.kent", "diana.prince", "peter.parker"], 
                  "yahoo.com": ["barbara.gordon", "jean.grey"], "hotmail.com": ["bruce.wayne"]}

for domain, users in dic.items():
    for user in users:
        print("{}@{}".format(user, domain))

clark.kent@gmail.com
diana.prince@gmail.com
peter.parker@gmail.com
barbara.gordon@yahoo.com
jean.grey@yahoo.com
bruce.wayne@hotmail.com


# Q2

The groups_per_user function receives a dictionary, which contains group names with the list of users. Users can belong to multiple groups. Fill in the blanks to return a dictionary with the users as keys and a list of their groups as values.



In [522]:
def groups_per_user(group_dictionary):
    user_groups = {}
    # Go through group_dictionary
    for groups, users in group_dictionary.items():
        # Now go through the users in the group
        for user in users:
     # Now add the group to the the list of
            # groups for this user, creating the entry
            # in the dictionary if necessary
            if user in user_groups:
                user_groups[user].append(groups)
            else:
                user_groups[user]=[groups]
    return(user_groups)

print(groups_per_user({"local": ["admin", "userA"],
		"public":  ["admin", "userB"],
		"administrator": ["admin"] }))

{'admin': ['local', 'public', 'administrator'], 'userA': ['local'], 'userB': ['public']}


In [578]:
def user_per_group(user_dictionary):
    group_users = {}
    for users, groups in user_dictionary.items():
        for group in groups:
            if group in group_users:
                group_users[group].append(users)
            else:
                group_users[group] = [users]
    return group_users

print(user_per_group({'admin': ['local', 'public', 'administrator'], 
                      'userA': ['local'], 'userB': ['public']}))

{'local': ['admin', 'userA'], 'public': ['admin', 'userB'], 'administrator': ['admin']}


# Q3
The dict.update method updates one dictionary with the items coming from the other dictionary, so that existing entries are replaced and new entries are added. What is the content of the dictionary “wardrobe“ at the end of the following code?



In [583]:
wardrobe = {'shirt': ['red', 'blue', 'white'], 'jeans': ['blue', 'black']}
new_items = {'jeans': ['white'], 'scarf': ['yellow'], 'socks': ['black', 'brown']}
wardrobe.update(new_items)
print(wardrobe)

{'shirt': ['red', 'blue', 'white'], 'jeans': ['white'], 'scarf': ['yellow'], 'socks': ['black', 'brown']}


# Q4
What’s a major advantage of using dictionaries over lists?


1. Dictionaries are ordered sets

2. Dictionaries can be accessed by the index number of the element

3. Elements can be removed and inserted into dictionaries

4. It’s quicker and easier to find a specific element in a dictionary



# Q5

The add_prices function returns the total price of all of the groceries in the dictionary. Fill in the blanks to complete this function.



In [588]:
def add_prices(basket):
    total = 0
    for product, prices in basket.items():
        for price in [prices]:
            total += price
#    return ("{:.2f}".format(total))
    return round (total, 2)


groceries = {"bananas": 1.56, "apples": 2.50, "oranges": 0.99, "bread": 4.59, 
             "coffee": 6.99, "milk": 3.39, "eggs": 2.98, "cheese": 5.44}
print(add_prices(groceries)) # Should print 28.44

28.44


In [590]:
def add_prices(basket):
	# Initialize the variable that will be used for the calculation
	total = 0
	# Iterate through the dictionary items
	for price in basket.values():
		# Add each price to the total calculation
		# Hint: how do you access the values of
		# dictionary items?
		total += price
	# Limit the return value to 2 decimal places
	return round(total, 2)  

groceries = {"bananas": 1.56, "apples": 2.50, "oranges": 0.99, "bread": 4.59, 
	"coffee": 6.99, "milk": 3.39, "eggs": 2.98, "cheese": 5.44}

print(add_prices(groceries)) # Should print 28.44

28.44


# Module Review

1. Basic structur wrap up
    - Strings
    - Lists
    - Dictionaries 


2. Data types
    - Tuples
    - Sets

# Module 4 Graded Assessment

# Q1 
The format_address function separates out parts of the address string into new strings: house_number and street_name, and returns: "house number X on street named Y". The format of the input string is: numeric house number, followed by the street name which may contain numbers, but never by themselves, and could be several words long. For example, "123 Main Street", "1001 1st Ave", or "55 North Center Drive". Fill in the gaps to complete this function.




In [677]:
def format_address(address_string):
    house_numer = ''
    street_name = ''
    spi = address_string.split()
    for element in spi:
        #if element.isnumeric():
        if element.isdigit():
            house_numer = element
        else:
            street_name += element
            street_name += " "
    return ("house number {} on street named {}".format(house_numer, street_name))

print(format_address("123 Main Street"))
# Should print: "house number 123 on street named Main Street"

print(format_address("1001 1st Ave"))
# Should print: "house number 1001 on street named 1st Ave"

print(format_address("55 North Center Drive"))
# Should print "house number 55 on street named North Center Drive"


house number 123 on street named Main Street 
house number 1001 on street named 1st Ave 
house number 55 on street named North Center Drive 


# Q2
The highlight_word function changes the given word in a sentence to its upper-case version. For example, highlight_word("Have a nice day", "nice") returns "Have a NICE day". Can you write this function in just one line?



In [688]:
def highlight_word(sentence, word):
    return(sentence.replace(word, word.upper()))
print(highlight_word("Have a nice day", "nice"))
print(highlight_word("Shhh, don't be so loud!", "loud"))
print(highlight_word("Automating with Python is fun", "fun")) 

Have a NICE day
Shhh, don't be so LOUD!
Automating with Python is FUN


# Q3
A professor with two assistants, Jamie and Drew, wants an attendance list of the students, in the order that they arrived in the classroom. Drew was the first one to note which students arrived, and then Jamie took over. After the class, they each entered their lists into the computer and emailed them to the professor, who needs to combine them into one, in the order of each student's arrival. Jamie emailed a follow-up, saying that her list is in reverse order. Complete the steps to combine them into one list as follows: **the contents of Drew's list, followed by Jamie's list in reverse order**, to get an accurate list of the students as they arrived.



In [733]:
def combine_lists(list1, list2):
    # Generate a new list containing the elements of list2
    new_list = list2
    # Followed by the elements of list1 in reverse order
    for i in reversed(range(len(list1))):
        new_list.append(list1[i]) 
    return new_list

Jamies_list = ["Alice", "Cindy", "Bobby", "Jan", "Peter"]
Drews_list = ["Mike", "Carol", "Greg", "Marcia"]

print(combine_lists(Jamies_list, Drews_list))


['Mike', 'Carol', 'Greg', 'Marcia', 'Peter', 'Jan', 'Bobby', 'Cindy', 'Alice']


In [740]:
def combine_lists(list1, list2):
    new_list = list2 + list1[::-1]
    return new_list
Jamies_list = ["Alice", "Cindy", "Bobby", "Jan", "Peter"]
Drews_list = ["Mike", "Carol", "Greg", "Marcia"]

print(combine_lists(Jamies_list, Drews_list))


['Mike', 'Carol', 'Greg', 'Marcia', 'Peter', 'Jan', 'Bobby', 'Cindy', 'Alice']


# Q4
Use a list comprehension to create a list of squared numbers (n*n). The function receives the variables start and end, and returns a list of squares of consecutive numbers between start and end inclusively. For example, squares(2, 3) should return [4, 9].



In [769]:
def squares(start, end):
    squared_number = []
    for x in range(start, end+1):
        squared_number.append(x*x)
    return squared_number

print(squares(2, 3)) # Should be [4, 9]
print(squares(1, 5)) # Should be [1, 4, 9, 16, 25]
print(squares(0, 10)) # Should be [0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

[4, 9]
[1, 4, 9, 16, 25]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [770]:
def squares(start, end):
    return [x * x for x in range(start, end+1)]
print(squares(2, 3)) # Should be [4, 9]
print(squares(1, 5)) # Should be [1, 4, 9, 16, 25]
print(squares(0, 10)) # Should be [0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

[4, 9]
[1, 4, 9, 16, 25]
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


# Q5
Complete the code to iterate through the keys and values of the car_prices dictionary, printing out some information about each one.



In [768]:
def car_listing(car_prices):
    result = ""
    for key, value in car_prices.items():
        result += "{} costs {} dollars".format(key, value) + "\n"
    return result

print(car_listing({"Kia Soul":19000, "Lamborghini Diablo":55000, "Ford Fiesta":13000, "Toyota Prius":24000}))

Kia Soul costs 19000 dollars
Lamborghini Diablo costs 55000 dollars
Ford Fiesta costs 13000 dollars
Toyota Prius costs 24000 dollars



# Q6
Taylor and Rory are hosting a party. They sent out invitations, and each one collected responses into dictionaries, with names of their friends and how many guests each friend is bringing. Each dictionary is a partial list, but Rory's list has more current information about the number of guests. Fill in the blanks to combine both dictionaries into one, with each friend listed only once, and the number of guests from Rory's dictionary taking precedence, if a name is included in both dictionaries. Then print the resulting dictionary.



In [803]:
def combine_guests(guests1, guests2):
    # Combine both dictionaries into one, with each key listed
    result = guests1
    for guesat in guests2:
        if guesat in  result:
            pass
        else:
            result[guesat] = guests2[guesat]  
    return result
                
  # only once, and the value from guests1 taking precedence

Rorys_guests = { "Adam":2, "Brenda":3, "David":1, "Jose":3, "Charlotte":2, "Terry":1, "Robert":4}
Taylors_guests = { "David":4, "Nancy":1, "Robert":2, "Adam":1, "Samantha":3, "Chris":5}
print(combine_guests(Rorys_guests, Taylors_guests))


{'Adam': 2, 'Brenda': 3, 'David': 1, 'Jose': 3, 'Charlotte': 2, 'Terry': 1, 'Robert': 4, 'Nancy': 1, 'Samantha': 3, 'Chris': 5}


In [807]:
from copy import deepcopy

def combine_guests(guests1, guests2):
    backup = deepcopy(guests1)
    guests1.update(guests2)
    for guest in guests1:
        if guest in backup:
            guests1[guest] = backup[guest]
    return guests1

Rorys_guests = { "Adam":2, "Brenda":3, "David":1, "Jose":3, "Charlotte":2, "Terry":1, "Robert":4}
Taylors_guests = { "David":4, "Nancy":1, "Robert":2, "Adam":1, "Samantha":3, "Chris":5}

print(combine_guests(Rorys_guests, Taylors_guests))

{'Adam': 2, 'Brenda': 3, 'David': 1, 'Jose': 3, 'Charlotte': 2, 'Terry': 1, 'Robert': 4, 'Nancy': 1, 'Samantha': 3, 'Chris': 5}


In [816]:
def combine_guests(guests1, guests2):
  # Combine both dictionaries into one, with each key listed 
  # only once, and the value from guests1 taking precedence
    guests2.update(guests1)
    return guests2

Rorys_guests = { "Adam":2, "Brenda":3, "David":1, "Jose":3, "Charlotte":2, "Terry":1, "Robert":4}
Taylors_guests = { "David":4, "Nancy":1, "Robert":2, "Adam":1, "Samantha":3, "Chris":5}

print(combine_guests(Rorys_guests, Taylors_guests))

{'David': 1, 'Nancy': 1, 'Robert': 4, 'Adam': 2, 'Samantha': 3, 'Chris': 5, 'Brenda': 3, 'Jose': 3, 'Charlotte': 2, 'Terry': 1}


# Q7
Use a dictionary to count the frequency of letters in the input string. Only letters should be counted, not blank spaces, numbers, or punctuation. Upper case should be considered the same as lower case. For example, `count_letters("This is a sentence.")` should return `{'t': 2, 'h': 1, 'i': 2, 's': 3, 'a': 1, 'e': 3, 'n': 2, 'c': 1}`.



In [873]:
def count_letters(text):
    text = text.lower()
    result = {}
    # Go through each letter in the text
    for letter in text:
        # Check if the letter needs to be counted or not
        if letter.isalpha():
            if letter in result:
                # Add or increment the value in the dictionary
                result[letter] += 1
            else:
                result[letter] = 1
    return result
print(count_letters("AaBbCc"))
# Should be {'a': 2, 'b': 2, 'c': 2}

print(count_letters("Math is fun! 2+2=4"))
# # # Should be {'m': 1, 'a': 1, 't': 1, 'h': 1, 'i': 1, 's': 1, 'f': 1, 'u': 1, 'n': 1}

print(count_letters("This is a sentence."))
# # # Should be {'t': 2, 'h': 1, 'i': 2, 's': 3, 'a': 1, 'e': 3, 'n': 2, 'c': 1}

{'a': 2, 'b': 2, 'c': 2}
{'m': 1, 'a': 1, 't': 1, 'h': 1, 'i': 1, 's': 1, 'f': 1, 'u': 1, 'n': 1}
{'t': 2, 'h': 1, 'i': 2, 's': 3, 'a': 1, 'e': 3, 'n': 2, 'c': 1}


# Q9
What do the following commands return when animal = "Hippopotamus"?



In [875]:
animal= 'Hippopotamus'
print(animal[3:6])
print(animal[-5])
print(animal[10:])

pop
t
us


# Q9
What does the list "colors" contain after these commands are executed?



In [878]:
colors = ["red", "white", "blue"]
colors.insert(2, "yellow")
print(colors)

['red', 'white', 'yellow', 'blue']


# Q10
What do the following commands return?



In [879]:
host_addresses = {"router": "192.168.1.1", "localhost": "127.0.0.1", "google": "8.8.8.8"}
host_addresses.keys()

dict_keys(['router', 'localhost', 'google'])